### PR TITLE
🚧 [BUU] Add `Edit` link into a small menu on the last Actions column to the right of the table

### LINKED ISSUE #11069: [BUU] Fully update the details of my products and variants
## Context
This issue introduces a new column in the table with actions. For this card the only action would be linking to the full edit product and variant pages.

## Out of scope
https://github.com/openfoodfoundation/openfoodnetwork/issues/11067
https://github.com/openfoodfoundation/openfoodnetwork/issues/11068

## Description

**- As an:** enterprise user 
**- On page:** `/admin/products`
**- I want to be able to:** edit product and variant data that's not in the table
**- So that:** I can update every detail of my catalogue

## Acceptance Criteria & Tests

**Scenario 1: New column**
Given that I am in `/admin/products`
And I have products in the list
When the list loads successfully
Then I see a column with the actions for my products and variants

**Scenario 2: Actions - product**
Given that I am seeing products in my catalogue
When I select the actions available for a product
Then I see an option to edit

**Scenario 3: Product - edit**
Given that I am seeing the actions for a product
When I select the edit option
Then I go to `/admin/products/[product-id]/edit`

**Scenario 4: Actions - Variant**
Given that I am seeing variants in my catalogue
When I select the actions available for a variant
Then I see an option to edit

**Scenario 5: Variant - edit**
Given that I am seeing the actions for a variant
When I select the edit option
Then I go to `admin/products/[product-id]/variants/[variant-id]/edit)`

## Design specs

Figma screens are available here > 
https://www.figma.com/file/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery---Back-Office?type=design&node-id=509%3A226&mode=design&t=PzBG8cWoBQnyy30t-1

Prototype here (open dropdown in first product) > https://www.figma.com/proto/v1zbrWDZSRd3Nqoe0SJ2Sm/Engineering-Delivery---Back-Office?page-id=509%3A226&type=design&node-id=509-227&viewport=655%2C222%2C0.36&t=dP5bR9HCNXI7CblU-1&scaling=min-zoom&mode=design

**New components and styles**

| Component | Example |
| --- | --- |
| Actions dropdown |Screenshot 2023-08-02 at 9 07 53|

### PR DESCRIPTION
**Pull request by jibees**
#### What? Why?

- Closes #11069

(there is no delete action in this PR, but I wanted to show here that I've prepared the CSS for a `.delete` item)

#### What should we test?

- As a hub manager, with feature toggle activated, check the new product page, and see that you can use that small component to fully edit product/variant (by going into its edition page)

#### Release notes

Changelog Category (reviewers may add a label for the release notes):

- [ ] User facing changes
- [ ] API changes (V0, V1, DFC or Webhook)
- [ ] Technical changes only
- [x] Feature toggled

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Cool.

File: app/webpacker/controllers/vertical_ellipsis_menu_controller.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+export default class extends Controller {
+  static targets = ["content"];
+
+  connect() {
+    super.connect();
+    window.addEventListener("click", (e) => {
+      if (this.element.contains(e.target)) return;
+      this.#hide();
+    });
```
Comment (dacook) [inline app/webpacker/controllers/vertical_ellipsis_menu_controller.js:11]: 🤖 Rubocop would say to have an empty line after a guard clause  ;)

```suggestion
    window.addEventListener("click", (e) => {
      if (this.element.contains(e.target)) return;

      this.#hide();
    });
```

Hmm I'd also suggest we don't embed functions for event listeners. Moving it to another function gives us a chance to make it easy to read:


```suggestion
    window.addEventListener("click", this.#hideIfClickedOutside);
```

File: spec/javascripts/stimulus/vertical_ellipsis_menu_controller_test.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    expect(content.classList.contains("show")).toBe(false);
+    button.click();
+    expect(content.classList.contains("show")).toBe(true);
+  });
+
+
+  it("remove show class from content when clicking button", () => {
+    const button = document.getElementById("button");
+    const content = document.getElementById("content");
```
Comment (dacook) [inline spec/javascripts/stimulus/vertical_ellipsis_menu_controller_test.js:37]: Gaetan taught me that these variables can be assigned in beforeEach to remove duplication (similar to rspec's `let!`).
https://github.com/openfoodfoundation/openfoodnetwork/pull/11509#discussion_r1328441837

In fact, I found I didn't even need the initial `let` that he mentioned. Just moving it to beforeEach worked.
https://github.com/openfoodfoundation/openfoodnetwork/pull/11509/files#diff-e2b6d56ae1511dc914a8a816c75e5a707679378ecd28340d001adf25fde5a468R48-R73

File: spec/system/admin/products_v3/products_spec.rb [hunk 3]
```diff
@@ -157,6 +157,36 @@
...
+    let!(:product_a) { create(:simple_product, name: "Apples", sku: "APL-00") }
+
+    before do
+      visit admin_products_v3_index_url
+    end
+
+    it "shows an actions memu with an edit link when clicking on icon for product" do
+      within row_containing_name("Apples") do
+        page.find(".vertical-ellipsis-menu").click
+        expect(page).to have_link "Edit", href: spree.edit_admin_product_path(product_a)
```
Comment (dacook) [inline spec/system/admin/products_v3/products_spec.rb:177]: I was thinking perhaps the test should actually click on the link to verify that it works. I guess it's possible that some random Javascript could `preventDefault` and prevent this link from working. 
But this seems sufficient, and much more efficient 👍

**Review by dacook — APPROVED** [review]
Looks great! 

The only thing I'd say is that I'd love to see it as a ViewComponent. You may have already thought this too. I think it almost has all the ingredients:
* general use Stimulus controller
* styles associated
* a partial.. but it's not general

I'm guessing it just needs to accept an array of arrays, like `[[label, url, type]]`. 
Perhaps for when you implement the delete button ;)

**Comment by jibees** [comment]
> You may have already thought this too. 

Yes, I thought about it too. Will do today!

File: app/webpacker/controllers/vertical_ellipsis_menu_controller.js (hunk 1 above)
Comment (jibees) [reply app/webpacker/controllers/vertical_ellipsis_menu_controller.js:11]: I agree. Will do.

File: spec/javascripts/stimulus/vertical_ellipsis_menu_controller_test.js (hunk 2 above)
Comment (jibees) [reply spec/javascripts/stimulus/vertical_ellipsis_menu_controller_test.js:37]: Oh yeah, much more easier. Super strange about this `const` assigned in another block but still available, but anyway...

Comment (dacook) [reply spec/javascripts/stimulus/vertical_ellipsis_menu_controller_test.js:37]: Yeah I don't understand, but am glad I don't need to 🙈

**Review by dacook — APPROVED** [review]
⚡ 🧹 👍

**Pull request merged**